server/libs/infra/src/search/typesense.repository.ts

```diff
@@ -0,0 +1,181 @@
+  numRetries: 3,
+  connectionTimeoutSeconds: 10,
+});
+
+@Injectable()
+export class TypesenseRepository implements ISearchRepository {
+  private client = typesense;
+  private logger = new Logger(TypesenseRepository.name);
+  private queue: Record<SearchCollection, any[]> = {
+    [SearchCollection.ASSETS]: [],
+    [SearchCollection.ALBUMS]: [],
+  };
+
+  constructor() {
+    setInterval(() => this.flush(), 5_000);
+  }
+
+  async setup(): Promise<void> {
+    // upsert collections
+    for (const [collectionName, schema] of schemas) {
+      const collection = await this.client
+        .collections(schema.name)
+        .retrieve()
+        .catch(() => null);
+      if (!collection) {
+        this.logger.log(`Creating schema: ${collectionName}/${schema.name}`);
+        await this.client.collections().create(schema);
+      } else {
+        this.logger.log(`Schema up to date: ${collectionName}/${schema.name}`);
+      }
+    }
+  }
+
+  async checkMigrationStatus(): Promise<SearchCollectionIndexStatus> {
+    const migrationMap: SearchCollectionIndexStatus = {
+      [SearchCollection.ASSETS]: false,
+      [SearchCollection.ALBUMS]: false,
+    };
+
+    // check if alias is using the current schema
+    const { aliases } = await this.client.aliases().retrieve();
+    this.logger.log(`Alias mapping: ${JSON.stringify(aliases)}`);
+
+    for (const [aliasName, schema] of schemas) {
+      const match = aliases.find((alias) => alias.name === aliasName);
+      if (!match || match.collection_name !== schema.name) {
+        migrationMap[aliasName] = true;
+      }
+    }
+
+    this.logger.log(`Collections needing migration: ${JSON.stringify(migrationMap)}`);
+
+    return migrationMap;
+  }
+
+  async index(collection: SearchCollection, item: AssetEntity | AlbumEntity, immediate?: boolean): Promise<void> {
+    const schema = schemaMap[collection];
+
+    if (immediate) {
+      await this.client.collections(schema.name).documents().upsert(item);
+      return;
+    }
+
+    this.queue[collection].push(item);
+  }
+
+  async import(collection: SearchCollection, items: AssetEntity[] | AlbumEntity[], done: boolean): Promise<void> {
+    try {
+      const schema = schemaMap[collection];
+      await this.client.collections(schema.name).documents().import(items, { action: 'upsert' });
+      if (done) {
+        await this.updateAlias(collection);
+      }
+    } catch (error: any) {
+      this.logger.error('Unable to index asset documents', error.importResults);
+      throw error;
+    }
+  }
+
+  search(collection: SearchCollection.ASSETS, query: string, filter: SearchFilter): Promise<SearchResult<AssetEntity>>;
+  search(collection: SearchCollection.ALBUMS, query: string, filter: SearchFilter): Promise<SearchResult<AlbumEntity>>;
+  async search(collection: SearchCollection, query: string, filters: SearchFilter) {
+    const alias = await this.client.aliases(collection).retrieve();
+
+    const { userId } = filters;
```

Comment:
Should we special-case userId as a separate function parameter to enforce it always being passed?

Reply:
It can still be required, even in this filters option. Thought was too have search service and it from the auth user object. It's business logic in my mind that typesense doesn't need to know about. Still haven't finalized the interface here yet.